src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerableFactory.cs

```diff
@@ -78,6 +79,8 @@ private static bool MatchesPattern(string expression, ReadOnlySpan<char> name, E
                     return FileSystemName.MatchesSimpleExpression(expression, name, ignoreCase);
                 case MatchType.Win32:
                     return FileSystemName.MatchesWin32Expression(expression, name, ignoreCase);
+                case MatchType.Regex:
+                    return FileSystemName.MatchesRegularExpressionString(expression, name.ToString(), ignoreCase);
```

Comment:
This pattern is very unfriendly to tree-shaking. It is close to impossible for the treeshaker to figure out that this is dead code. It means that anybody using this file enumeration will now get the regex engine, including its transitive closure.

Reply:
Any suggestions? I'm afraid I have no idea how tree-shaking analyzes things.

Reply:
Is calling through a delegate gotten from reflection, too slow?

Reply:
It think that introducing a static dependency from basic FileSystem contract on regex engine is a bad idea. 

> Is calling through a delegate gotten from reflection, too slow?

I agree. This API should be designed as composable - if you need full regex matching, you should supply it as a delegate when constructing the matcher.

Reply:
> I agree. This API should be designed as composable - if you need full regex matching, you should supply it as a delegate when constructing the matcher.

I'm not sure I follow. Are you suggesting that I grab Regex.IsMatch internally via reflection in FileSystemName?

Reply:
I am suggesting that System.IO.FileSystem knows nothing about RegEx. If somebody wants regex matching, they create regex matcher themselves and supply it as callback to System.IO.FileSystem.

Reply:
> If somebody wants regex matching, they create regex matcher themselves and supply it as callback to System.IO.FileSystem.

Is there any way to structure this so that tree shakers can figure out usage? Having our existing APIs support regex style matching without contortions on the developer end is desirable, particularly for the point of getting performance improvements automatically from new versions of CoreFX.

Reply:
> Is there any way to structure this so that tree shakers can figure out usage?

Nothing very reasonable. We have similar legacy design in System.Net.WebRequest where you always get mailto protocol support in the transitive close even if you just want to talk over http. Everybody agrees that it is pretty dump design that we should not repeat.

> particularly for the point of getting performance improvements automatically from new versions of CoreFX

I do not understand this what this has to do with performance. I think it is quite the opposite: I expect that we will eventually get compile time regex built into the language (because of every other language is getting it). Once that happens the composable pattern can be optimized at compile time (no need to waste time creating the regex dynamically at runtime). The enum based regex pattern that you are proposing will be impossible to optimize this way.

Reply:
>> Is there any way to structure this so that tree shakers can figure out usage?

> Nothing very reasonable.

Really? `MatchType` can be anything, it doesn't have to be an enum. Even if it's crazy I'd like to understand possible patterns so I can think in terms of the tree-shaker in the future.

Outside of goofy internal contortions we could turn `MatchType` into an interface or abstract class. `FileExpressionEvaluator` maybe? Then provide a matcher from Regex (which would put a dependency on FileSystem) or add a S.IO.FS.Regex?

Or possibly provide `MatchType.Custom` and have an additional member that holds an evaluator?

How about creating a `RegexEnumerable<TResult> : FileSystemEnumerable<TResult>`? That wouldn't allow using through `Directory.GetFiles` or anything, but would at least allow us to improve the implementation details in the future.

> I do not understand this what this has to do with performance.

Point is that we could switch to some new API such as `Regex.IsMatch(ReadOnlySpan<char>, ReadOnlySpan<char>)` or some other optimized implementation in a future release.

Reply:
> MatchType can be anything

I means nothing reasonable - as long as the matchtype is enum. Can the MatchType be just a delegate?

> `Regex.IsMatch(ReadOnlySpan<char>, ReadOnlySpan<char>)` or some other optimized implementation in a future release

If we want to shoehorn APIs in last minute for 2.1, it should rather be Regex matcher that takes Spans.

Reply:
> Can the MatchType be just a delegate?

Yeah, either that or have an option that allows you to pass a delegate.

> it should rather be Regex matcher that takes Spans.

One of the other things that we might be able to do in the future if we provide a matcher Regex in S.IO.FS is that we could swap in a UTF-8 implementation on the Unix side. It would be nice to be able to continually roll in internal smarts without exposing everything. While I'm entertaining exposing very platform specific data in the enumeration APIs (e.g. the native UTF-8 bytes) ideally we wouldn't have to but for the most obscure of scenarios.

Reply:
> If we want to shoehorn APIs in last minute for 2.1, it should rather be Regex matcher that takes Spans

We don't have time to implement a real one but we could add one that immediately converted to string; folks could use it now, and we make efficient next release. Is that what you mean? It could be an antipattern if callers had a strong to give us.

Reply:
I agree. Maybe we will want overloads that take an output buffer also and I don't want a profusion of overloads.

Reply:
> If it was my call, I would just postpone it to the next version.

+1